docs/rfcs/1005-realm-key-rotation.md

```diff
@@ -0,0 +1,821 @@
+
+> **TODO:**
+>
+> The fields `encryption_algorithm` & `hash_algorithm` allows for future evolution.
+> The idea would be to switch to  Chacha20-Poly1305 given it is a more scrutinized
+> algorithm with audited implementations.
+>
+> Another benefit is to be able to easily upgrade those algorithm in the future in
+> case of exploit in them.
+
+### 4.4 New document `RealmKeysBundle`
+
+```json5
+{
+    "label": "RealmKeysBundle",
+    "type": "realm_keys_bundle",
+    "other_fields": [
+        {
+            "name": "author",
+            "type": "DeviceID",
+        },
+        {
+            // Timestamp must be equal to the one in the `RealmRoleCertificate`
+            "name": "timestamp",
+            "type": "DateTime"
+        },
+        {
+            "name": "realm",
+            "type": "VlobID"
+        },
+        {
+            "name": "keys",
+            // Keys ordered by index
+            "type": "List<Bytes>"
+        }
+    ]
+}
+```
+
+### 4.5 New document `RealmKeysBundleAccess`
+
+```json5
+{
+    "label": "RealmKeysBundleAccess",
+    "type": "realm_keys_bundle_access",
+    "other_fields": [
+        {
+            "name": "author",
+            "type": "DeviceID"
+        },
+        {
+            // Timestamp must be equal to the one in the `RealmRoleCertificate`
+            // or `RealmKeyRotationCertificate` (depending if the access has been
+            // given before or after last key rotation)
+            "name": "timestamp",
+            "type": "DateTime"
+        },
+        {
+            "name": "realm",
+            "type": "VlobID"
+        },
+        {
+            "name": "key_index",
+            "type": "IndexInt"
+        },
+        {
+            "name": "keys_bundle_key",
+            "type": "SecretKey"
+        },
+        {
+            // Only present when the `RealmKeysBundleAccess` is generated for use with
+            // `realm_share` command (and not `realm_rotate_key` given in this case the
+            // user already )
+            "name": "workspace_label",
+            "type": "RequiredOption<String>"
+        }
+    ]
+}
+```
+
+> **TODO:**
+>
+> Do we need author/timestamp/realm_id in `KeysBundle`&`KeysBundleAccess` ?
+>
+> The need is to ensure they both correspond to a given `RealmKeyRotationCertificate`, but
+> adding those fields seems a lot of duplicated data :'(
+>
+> And we cannot simply add `KeyBundle`&`KeysBundleAccess` hashes to the certificate
+> given they contains encryption keys (and we cannot do HMAC).
+>
+> Another solution is to "go light":
+>
+> - `KeysBundleAccess` & `KeysBundle` are not signed (only encrypted)
+> - `KeysBundleAccess` only contains `keys_bundle_key` and `workspace_label`
+> - `KeysBundle` only contains the keys
+> - if `KeysBundleAccess` is forged then we cannot decrypt `KeysBundle`
+> - if `KeysBundle` is forged, then the key canary in the certificate doesn't work
+>
+> Do you see a flaw here ?
```

Comment:
TODO

Reply:
> The need is to ensure they both correspond to a given `RealmKeyRotationCertificate`, but
> adding those fields seems a lot of duplicated data :'(


Certificate are identified with a serial, we could do something similar with an UUID. That would removing the need to specify the fields `author&co` everytime but only a single field (`id|bundle_id|serial|bundle_serial` :shrug:).

Reply:
> Certificate are identified with a serial

Not sure what you mean, currently there is no serial number on the certificates (they have an index based on their order by timestamp)

> we could do something similar with an UUID. That would removing the need to specify the fields author&co everytime but only a single field (id|bundle_id|serial|bundle_serial 🤷).

The fields author&timestamp are useful to validate a signature. Here we can get those fields back from the related certificate, but it makes things more complex...
Another solution is to only encrypt the keysbundle & keysbundleaccess, but then the bundle_id/bundle_serial values are not guaranteed to be the right one... unless we take into account the encryption layer (given we do AEAD), in this case we are back to the "go light" solution, do you agree ?

Reply:
we are good for the "go light" approach :+1: